### PR TITLE
Fix broken isStyledComponent type

### LINKED ISSUE #1875: Component is not defined before
By using this library, I realize that Component is not defined before and Component is called at 98 line as shown below in the link anchor, as you can see Component is imported at line 2 then an alias called ReactComponent is assigned to that, or maybe I'm wrong.

https://github.com/styled-components/styled-components/blob/26be229adbc5629927dead3b95cf1076d88abcd6/typings/styled-components.d.ts#L98

### PR DESCRIPTION
**Pull request by MayhemYDG**
Fixes #1875

### CODE REVIEW

**Comment by MayhemYDG** [comment]
Regression comes from this commit: https://github.com/styled-components/styled-components/commit/ef920d8b1b06ec6068fda3fbf74ac82221b054fc#diff-bd3406937467ea42646ca601d869aa62L4

**Comment by MayhemYDG** [comment]
@kitten I don't know enough about styled-components to make the right call here, but this is breaking our build so a fix is needed.

**Comment by Igorbek** [comment]
@kitten agree, that would be a better choice.

File: typings/styled-components.d.ts [hunk 1]
```diff
@@ -95,7 +95,7 @@ export function withTheme<P extends { theme?: T; }, T>(component: React.Componen
 export function keyframes(strings: TemplateStringsArray, ...interpolations: SimpleInterpolation[]): string;
 export function injectGlobal(strings: TemplateStringsArray, ...interpolations: SimpleInterpolation[]): void;
 export function consolidateStreamedStyles(): void;
-export function isStyledComponent(target: string | Component<object>): boolean;
+export function isStyledComponent(target: string | React.Component<object>): boolean;
```
Comment (Igorbek) [inline typings/styled-components.d.ts]: So please replace this line with:
```diff
-export function isStyledComponent(target: string | React.Component<object>): boolean;
+export function isStyledComponent(target: string | ReactComponent<object>): target is StyledComponentClass;
```

**Comment by probablyup** [comment]
Please also add a changelog entry!

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,7 +6,7 @@ _The format is based on [Keep a Changelog](http://keepachangelog.com/) and this
 
 ## Unreleased
 
-* placeholder
+* Fixed a bug in typings where `isStyledComponent` was defined using an undefined variable. (see #1876)
```
Comment (probablyup) [inline CHANGELOG.md]: Please follow the format

File: typings/styled-components.d.ts [hunk 3]
```diff
@@ -1,122 +1,206 @@
...
+  strings: TemplateStringsArray,
+  ...interpolations: SimpleInterpolation[]
+): string
+export function injectGlobal(
+  strings: TemplateStringsArray,
+  ...interpolations: SimpleInterpolation[]
+): void
+export function consolidateStreamedStyles(): void
+export function isStyledComponent(
+  target: string | ReactComponent | StyledComponentClass
```
Comment (probablyup) [inline typings/styled-components.d.ts]: this is the relevant change

```ts
target: string | ReactComponent | StyledComponentClass
```

**Comment by mxstbr** [comment]
Thank you so much for helping us improve styled-components! Based on our [Community Guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md) every person that has a PR of any kind merged is offered **an invitation to the styled-components organization**—that is you right now!

Should you accept, you'll get write access to the main repository. (and a fancy `styled-components` logo on your GitHub profile!) You'll be able to label and close issues, merge pull requests, create new branches, etc. We want you to help us out with those things, so don't be scared to do them! Make sure to read our [contribution and community guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md), which explains all of this in more detail. Of course, if you have any questions just let me know!

File: typings/styled-components.d.ts [hunk 4]
```diff
@@ -1,122 +1,208 @@
...
+): void
 
-export function keyframes(strings: TemplateStringsArray, ...interpolations: SimpleInterpolation[]): string;
-export function injectGlobal(strings: TemplateStringsArray, ...interpolations: SimpleInterpolation[]): void;
-export function consolidateStreamedStyles(): void;
-export function isStyledComponent(target: string | Component<object>): boolean;
+export function consolidateStreamedStyles(): void
 
-export const ThemeProvider: ThemeProviderComponent<object>;
+export function isStyledComponent(target: any): target is StyledComponentClass
```
Comment (Idered) [inline typings/styled-components.d.ts]: I got error on this line:

```
ERROR in [at-loader] ./node_modules/styled-components/typings/styled-components.d.ts:179:59    
TS2707: Generic type 'StyledComponentClass<P, T, O>' requires between 2 and 3 type arguments.
```

Using with awesome-typescript-loader and typescript 3.0.1

Comment (aprilrd) [reply typings/styled-components.d.ts]: I am using typescript 2.9.2 and getting the same error.

Comment (matepapp) [reply typings/styled-components.d.ts]: Same here.

Comment (PzYon) [reply typings/styled-components.d.ts]: Is fixed in PR #1885

Comment (Dangoo) [reply typings/styled-components.d.ts]: @PzYon But not present in the published version @2.4.1 on the npm :(
/cc @probablyup

**Pull request merged**